crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -61,55 +147,252 @@ pub(crate) struct Exchange<T> {
     sender_counters: Vec<CachePadded<AtomicUsize>>,
     /// Callback invoked when all `npeers` mailboxes are available.
     sender_callbacks: Vec<OnceCell<Box<dyn Fn() + Send + Sync>>>,
+    /// The number of workers that have already sent their messages in the
+    /// current round.
+    sent: AtomicUsize,
+    /// The RPC clients to contact remote hosts.
+    clients: Arc<Clients>,
+    /// For each sender/receiver worker pair involved in RPC, this allows the
+    /// `exchange` RPC to wait until the receiver has taken its data out of the
+    /// mailbox.
+    sender_notifies: Vec<Notify>,
+    /// A callback that takes the raw data exchanged over RPC and deserializes
+    /// and delivers it to the receiver's mailbox.
+    deliver: Box<dyn Fn(Vec<u8>, usize, usize) + Send + Sync + 'static>,
 }
 
-impl<T> Exchange<T>
-where
-    T: Send + 'static,
-{
-    /// Create a new exchange operator for `npeers` communicating threads.
-    fn new(npeers: usize) -> Self {
+impl InnerExchange {
+    fn new(
+        exchange_id: ExchangeId,
+        deliver: impl Fn(Vec<u8>, usize, usize) + Send + Sync + 'static,
+        clients: Arc<Clients>,
+    ) -> InnerExchange {
+        let runtime = Runtime::runtime().unwrap();
+        let npeers = runtime.num_workers();
         Self {
+            exchange_id,
             npeers,
-            mailboxes: (0..npeers * npeers).map(|_| Mutex::new(None)).collect(),
-            receiver_counters: (0..npeers)
-                .map(|_| CachePadded::new(AtomicUsize::new(0)))
-                .collect(),
+            local_workers: runtime.layout().local_workers(),
+            clients,
+            receiver_counters: (0..npeers).map(|_| AtomicUsize::new(0)).collect(),
             receiver_callbacks: (0..npeers).map(|_| OnceCell::new()).collect(),
+            sender_notifies: (0..npeers * npeers).map(|_| Notify::new()).collect(),
```

Comment:
`npeers * num_local_workers` should be enough?

Reply:
Even `n_local_workers * n_remote_workers`. I'll change this.